Sprint 4

sgp900 opened the pull request:
1. Сборщик Parcell заменён на Webpack;

2. В проект добавлена система тестирования на базе mocha и chai;

3. Написаны тесты для:
   3.1. роутера
   3.2. компонента link src/components/Link/link.test.ts
   3.3. компонента placeHolderInput src/components/PlaceHolderInput/placeHolderInput.test.ts
   3.4. страницы входа src/pages/Login/login.test.ts
   3.5. модуля отправки сообщений src/utils/HTTPTransport.test.ts
   3.6. базового блока src/utils/block.test.ts
   3.7. хэлпера src/utils/helpers.test.ts

4. С помощью Husky, настроен Git precommit.

5. Создан Docker контейнер и задеплоен на render.com

sgp900 committed:
1. Сборщик Parcell заменён на Webpack;

2. В проект добавлена система тестирования на базе mocha и chai;

3. Написаны тесты для:
3.1. роутера
3.2. компонента link src/components/Link/link.test.ts
3.3. компонента placeHolderInput src/components/PlaceHolderInput/placeHolderInput.test.ts
3.4. страницы входа  src/pages/Login/login.test.ts
3.5. модуля отправки сообщений src/utils/HTTPTransport.test.ts
3.6. базового блока src/utils/block.test.ts
3.7. хэлпера src/utils/helpers.test.ts

4. С помощью Husky, настроен Git precommit.

sgp900 committed:
Создан Docker контейнер и задеплоен на render.com
HTTP адрес внесен в README.md

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации, 

* в приложении не работают сокеты и ошибки в консоли https://disk.yandex.ru/i/a7T_cenmeaLIYg

gennady-bars commented on .husky/pre-commit:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+npm test
```
в `pre-commit` нужно еще линтеры запускать

gennady-bars reviewed (commented):
еще заметил ошибки:

* в `pre-commit` нужно еще линтеры запускать
*

sgp900 committed:
1. К хуку pre-comit В Husky добавлен запуск линтеров.
2. Чтобы отловить ошибку createObjectURL добавлена дополнительная проверка, на длину пути. А также, сам блок кода обернут в try/cath. К сожалению, локально не удалось воспроизвести ошибку.

gennady-bars commented on webpack.config.build.js:
```diff
@@ -0,0 +1,50 @@
+const HtmlWebpackPlugin = require("html-webpack-plugin");
+
+module.exports = {
+  mode: "development",
```
**Можно лучше**

`webpack.config.js`  хорошо бы разбить на несколько частей - `development` и `production`. Так же будет еще файл с общими частями для обеих сборок. Объединить конфиги можно с помощью библиотеки `webpack-merge`

gennady-bars commented on package.json:
```diff
@@ -5,30 +5,51 @@
   "scripts": {
     "csslint": "stylelint \"**/*.scss\"",
```
**Можно лучше**


Для линтинга на прекоммите можно воспользоваться библиотекой [lint-staged](https://www.npmjs.com/package/lint-staged) . Она позволяет осуществлять линтинг только тех файлов, которые **изменились**

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.




* `Readme` хорошо оформлен
* Проект задеплоен на `Render` и отлично работает
* Отлично, что фиксируете версии зависимостей
* Отлично, что нет `EOF` ошибок в гите
* Хорошая структура папок и файлов
* Отлично, что в `pre-commit` запускаются и тесты, и линтеры
* Все тесты (28 шт) проходят успешно. Хорошее кол-во тестов
* `npm audit` показывает **ноль** уязвимостей
*  Проект отлично выглядит и функционирует


Удачного дальнейшего обучения.


**Можно лучше**

* `webpack.config.js`  хорошо бы разбить на несколько частей - `development` и `production`. Так же будет еще файл с общими частями для обеих сборок. Объединить конфиги можно с помощью библиотеки `webpack-merge`
* Для линтинга на прекоммите можно воспользоваться библиотекой [lint-staged](https://www.npmjs.com/package/lint-staged) . Она позволяет осуществлять линтинг только тех файлов, которые **изменились**
* Попробуйте избавиться от типов `any`
